code/game/machinery/firealarm.dm

```diff
@@ -32,31 +30,49 @@
 
 	//Trick to get the glowing overlay visible from a distance
 	luminosity = 1
-
-	var/detecting = 1
-	var/buildstage = 2 // 2 = complete, 1 = no wires, 0 = circuit gone
-	COOLDOWN_DECLARE(last_alarm)
-	var/area/myarea = null
-	//Has this firealarm been triggered by its enviroment?
-	var/triggered = FALSE
+	///Buildstate for contruction steps. 2 = complete, 1 = no wires, 0 = circuit gone
+	var/buildstage = 2
+	///Our home area, set in Init. Due to loading step order, this seems to be null very early in the server setup process, which is why some procs use `my_area?` for var or list checks.
+	var/area/my_area = null
+	///looping sound datum for our fire alarm siren.
+	var/datum/looping_sound/firealarm/soundloop
 
 /obj/machinery/firealarm/Initialize(mapload, dir, building)
 	. = ..()
 	if(building)
 		buildstage = 0
 		panel_open = TRUE
 	update_appearance()
-	myarea = get_area(src)
-	LAZYADD(myarea.firealarms, src)
+	my_area = get_area(src)
+	LAZYADD(my_area.firealarms, src)
 
 	AddElement(/datum/element/atmos_sensitive, mapload)
 	RegisterSignal(SSsecurity_level, COMSIG_SECURITY_LEVEL_CHANGED, .proc/check_security_level)
+	soundloop = new(src, FALSE)
 
 /obj/machinery/firealarm/Destroy()
-	myarea.firereset(src)
-	LAZYREMOVE(myarea.firealarms, src)
+	if(my_area)
+		LAZYREMOVE(my_area.firealarms, src)
+		my_area = null
+	QDEL_NULL(soundloop)
 	return ..()
 
+/**
+ * Sets the sound state, and then calls update_icon()
+ *
+ * This proc exists to be called by areas and firelocks
+ * so that it may update its icon and start or stop playing
+ * the alarm sound based on the state of an area variable.
+ */
+/obj/machinery/firealarm/proc/set_status()
+	if( (my_area.fire || LAZYLEN(my_area.active_firelocks)) && !(obj_flags & EMAGGED) )
+		soundloop.start()
+		set_light(l_power = 0.8)
+	else
+		soundloop.stop()
+		set_light(l_power = 0)
+	update_icon()
```

Comment:
I wonder if it would be better to do the sound stuff based on the fire alarm's own state (triggered, in a breach/fire), rather then the area's, so it doesn't layer horribly

Reply:
For what it's worth, I cut the sound file down to two beeps, but left the loop time what it was before. It's now about 75% silence and is way better even if you have different fire alarms nearby on different loop timers.